keps/4136-admission-fair-sharing/README.md

```diff
@@ -0,0 +1,245 @@
+# KEP-4136: Admission Fair Sharing
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This KEP describes the mechanism for fair admission of workloads coming from a group of
+different sources (like Cluster and Local Queues) based on the source shared resource usage.
+Workloads from sources that use less are admitted before workloads coming from sources that use more. 
+
+## Motivation
+
+Currently Kueue has a Fair Sharing mechanism that enforces fair-sharing of unused resources
+via preemption. If one Cluster Queue is using much more of the resources than the other one
+that is in need, some workloads from the first one may be preempted to allow more “fair” distribution. 
+
+This model has multiple assumptions:
+
+* Workloads can be preempted.
+* Users get a proper quota to do their regular business.
+* Users come from multiple teams/organizations and they would rather have a strict policy but 
+fair policy than show understanding to those consuming all shared resources. 
+
+However, these assumptions are not universal. Sometimes:
+
+* Workloads should not be preempted.
+* Users operate on a shared but bigger quota.
+* Fairness should not win over getting the workloads eventually completed.
+
+In that case the existing mode doesn’t work and a different one needs to be employed.
+
+### Goals
+
+* Establish a method for how shared resource usage is calculated and recorded and how users can fine tune the mechanism.
+* Allow to specify a fair admission scope at either individual Cluster Queue or Cohort scope.
+* Allow to specify the relative importance of LocalQueues targeting the same ClusterQueue.
+* Amend the admission mechanism to work on admission scopes instead of only on ClusterQueues. 
+* Select the appropriate admission candidates for each of the admission scopes and admit them according to the selected queueing policy.
+* Make the new mechanism complementary to the existing preemption-based fair sharing
+
+### Non-Goals
+
+* Store time series data inside K8S.
+* Provide precise shared resource usage accounting or billing.
+
+## Proposal
```

Comment:
Could you expand a bit more and clarify the status of preemption and reclamation. In particular, if the new fair sharing mode supports priority based preemption, or reclamation? So, if it supports reclamation, how are we going to select the candidates: as in the "classical preemption", the "classical fair sharing" , or would there be a new algo?

Reply:
So even both fair sharing approaches can work at the same time for a cohort, ie. The new fair sharing will order workloads for scheduling, and the old fair sharing will determine reclamation targets. Right?

Reply:
This makes sense, thanks for clarifying. 

Still, the clash of naming is confusing, as these are not just two modes of the same.

So I'm thinking about renaming the feature to avoid the collision. I guess the main "selling point" about the feature is being past usage aware, and making "fair" admission choices in that regard. 
1. Fair Admission
2. Past-Aware Admission
3. Past Usage-Aware Admission
4. Past Usage-Aware Fair Sharing

I'm leaning to (3.) to emphasize that the "past usage" awareness is the main feature here. wdyt?

cc @gabesaba @PBundyra @pajakd if you have some other thoughts.

Reply:
Not sure how we should proceed with this, but I'm not the biggest fan of how extensive 2,3,4 are. Also not sure whether the name clash is a bad thing - we reflect those two modes also in API, both in config, and in objects' status

Reply:
>  I'm not the biggest fan of how extensive 2,3,4 are

You mean long? I think `Past Usage-Aware Admission` (my fav) isn't really much longer than others. 

> we reflect those two modes also in API, both in config, and in objects' status

Yeah, and I find this is confusing too. This is because this feature is not a new mode of "Fair Sharing" (as the name suggests). It can actually be used in combination with the current Fair Sharing as pointed out in https://github.com/kubernetes-sigs/kueue/pull/4252#discussion_r2045007906.

Reply:
Actually, after syncing with @mwielgus and @PBundyra I'm ok with the current naming. The argument is that the new feature is conceptually a new mode of fair sharing at the level of admission. The "Past Usage-Awareness" is an important aspect of the current implementation, but maybe another approach will be used in the future. 

Still, I believe the naming can be confusing here, but we can address it at the documentation level. Feel free to resolve.